src/Psalm/Internal/Analyzer/Statements/Expression/ArrayAnalyzer.php

```diff
@@ -139,19 +118,39 @@ public static function analyze(
             $item_value_type = null;
         }
 
+        // if this array looks like an object-like array, let's return that instead
+        if (!empty($array_creation_info->property_types)) {
+            $atomic_type = new TKeyedArray($array_creation_info->property_types, $array_creation_info->class_strings);
+            if ($array_creation_info->can_create_objectlike) {
+                $atomic_type->sealed = true;
+            } else {
+                $atomic_type->previous_key_type = $item_key_type ?? Type::getArrayKey();
+                $atomic_type->previous_value_type = $item_value_type ?? Type::getMixed();
+            }
+            $atomic_type->is_list = $array_creation_info->all_list;
+
+            $stmt_type = new Union([$atomic_type]);
+
+            if ($array_creation_info->parent_taint_nodes) {
+                $stmt_type->parent_nodes = $array_creation_info->parent_taint_nodes;
+            }
+
+            $statements_analyzer->node_data->setType($stmt, $stmt_type);
+
+            return true;
+        }
+
         if ($item_key_type === null && $item_value_type === null) {
             $statements_analyzer->node_data->setType($stmt, Type::getEmptyArray());
 
             return true;
         }
 
         if ($array_creation_info->all_list) {
-            if (empty($array_creation_info->item_key_atomic_types)) {
+            if ($array_creation_info->can_be_empty) {
                 $array_type = new TList($item_value_type ?? Type::getMixed());
             } else {
                 $array_type = new TNonEmptyList($item_value_type ?? Type::getMixed());
-                /** @psalm-suppress InvalidPropertyAssignmentValue */
-                $array_type->count = count($array_creation_info->property_types);
```

Comment:
This won't cause issues?

Reply:
It actually causes issues if it's there because `property_types` is guaranteed to be empty at that point since any array with `property_types` set will be turned into an array shape (I only noticed because Psalm warned me...). I'm also not convinced it wasn't buggy before, it seems like it could probably have been setting `count` even when it wasn't exact.